### ISSUE TITLE
Quartz Web Dashboard

### ISSUE THREAD

**Issue by lahma** [quartznet/quartznet#251]
There should be an embedded view which allows to see what's happening inside a process that hosts Quartz. Some initial thoughts:
- works as a plugin that can be attached to Quartz setup, not required by default
- uses Owin self hosting
- SignalR/WebSockets for live logs 
- REST like API for basic scheduler tasks, "remoting V2"
- minimal set of required libraries, but Owing+SignalR+WebAPI most likely will bloat a bit
- SPA application based on Aurelia
- basic managements tasks like triggering job available
- investigate support to create persisted job execution history, backed by database
- later possibility to get a "cluster view" that contains consolidated view from all registered scheduler processes

**Comment by jvilalta** [comment]
Marko, 
I think all of the above are necessary

I'd like to suggest also:
- the ability to add multiple schedulers so that you can easily switch between schedulers.
- some sort of authentication/authorization mechanism

The QuartzNetFeaturePack has a plugin that uses owin (nancy but that's because back then webapi didn't run outside of iis) and has a very simple api started. I just started yesterday on moving the front end to Aurelia, so I'm good with that as well. 

Let me know how I can help. Happy to move that into the main project or keep it separate for now.

J

**Comment by mgonzales3** [comment]
Marko,

what about a conmaker included

**Comment by lpmeunier** [comment]
<<REST like API for basic scheduler tasks, "remoting V2">>

Very much this, except replace basic with complete.

**Comment by Jawvig** [comment]
At the moment we've got a (far too basic) version of this embedded in our already existing admin app. I'd love to shift to a community supported version, but not too keen on standing up a separate admin app. As long as the data that it displays can be easily retrieved from a new REST API then I'd have the option of either using an out of the box UI or continuing to maintain a UI integrated with our existing app but sourcing the data from a REST API. The remoting API is a slightly hard to work with, especially when the scheduler is down.

In terms of features, supporting multiple schedulers would be good for us too. My plan had been to merge the data from both schedulers into a single view with an indication of which scheduler a job was running in.

Finally, I did write an Entity Framework 6 based persisted job execution history that I'd be happy to contribute. It's implemented as an IJobListener. It does have a slight dependency on Autofac at the minute just to get easy lifetime management (via the Owned concept) but this could be removed. Let me know if you're interested enough for me to package it up a bit to share it out.

**Comment by vgl** [comment]
Hi, where I can find a road map for 3.0 version, and when it will be available. Because right now I'm doing same - implementing web ui to handle quartz tasks...

**Comment by vgl** [comment]
Any dates for 3.0 version?

**Comment by lahma** [comment]
The current road map is roughly the issues listed in the tracker:

https://github.com/quartznet/quartznet/milestones/3.0

Some issues should then opened for discussion, just like this web dashboard one where we can discuss the individual requirements.

There's no date for 3.0, just the famous "it's done when it's done".  As there is no dedicated personnel in this project it's all about finding the free time to spare. So unfortunately I can't give estimates because we'd be probably failing them...

**Comment by vgl** [comment]
Does bootstrap v3 used on fronted side?

**Comment by vgl** [comment]
What about security layer for the dashboard and api controllers? Does it implemented or it just in plans for now?

**Comment by lahma** [comment]
Yes, the UI is based on bootstrap at the moment. A security layer probably should be implemented too.

**Comment by lahma** [comment]
I'm dropping this from 3.0 to deliver the basic functionality of the core at least. Haven't had the required time for this undertaking.

**Comment by Pinox** [comment]
Hi @lahma ,

Just want to know if the dashboard is still on the radar screen.  Been using qrtz with asp.net core but dashboard would be seriously useful. Something like https://github.com/HangfireIO/Hangfire

It's open source and something similiar in qrtz would be great, i don't think anyone needs to reinvent the wheel here.

**Comment by lahma** [comment]
@Pinox I would love such a feature too. Contributions are more than welcome to get things rolling - PRs targeting the web project adding functionality etc.

**Comment by pierluca** [comment]
As I was looking for this feature, I found my needs to be mostly covered by https://github.com/guryanovev/CrystalQuartz

**Comment by Pinox** [comment]
@pierluca  thanks bud, I did not know this.

**Comment by lahma** [comment]
I'd like to ask both @jlucansky and @guryanovev for an opinion. Would it make sense to develop shared JSON API on Quartz's side (see API related requirements in issue description) and then adapt both of your UI parts to share?

I know that you both have put hard work into your management UIs and they probably are opionated and follow the requirements you have set for them, both possible having pros and cons. So my guess is that there's neither an easy merge of UI solutions nor much requirement to do so (unless you see the benefit).

Easily accessible REST'ish API would allow users to define Yet Another UI / console application / integration they need, also covering .NET Core which now has none from Quartz's side. I already had some effort on building one but now that there are three competing solutions for the same goal of having HTTP exposed management methods it feels a bit silly.

**Comment by jlucansky** [comment]
I created Quartzmin with no intention to be used as the universal REST API so it is not as clean and straightforward as it should be. But, as part for my another work project, I was forced to replace .NET remoting with web API and I crafted pretty complete API for IScheduler. Right now, it targets only full .NET Framework but it can be used as starting point or inspiration: https://gist.github.com/jlucansky/1c5b8e4befa3b745b4485184811caa2f 

There is also working client side if you are interested...

**Comment by kosalanuwan** [comment]
Just thot of bringing this conversation to live. I've been using Quartz for some time now and having a Dashboard is a great deal. 

I came across Hangfire and Coravel while roaming in GitHub.

I'm not a fan of Hangfire's programming model, especially their DI wiring, and the use of schedules, tasks, queues, triggers et al is a bit wonky.

On the other hand, Coravel programming model is quite promising as they try to design the APIs Fluently and use .NET native APIs such as `IHostedService` but not quite sure whether they use `PeriodicTimer`, however, their Dashboard is only for Pro users and their latest release is nearly 2yo.

Sadly, both [@jlucansky/Quartzmin](https://github.com/jlucansky/Quartzmin) and [@guryanovev/CrystalQuartz](https://github.com/guryanovev/CrystalQuartz) have no activity for nearly a half-a-decade (4-5y) and .NET has changed a lot thus far.

/cc @lahma, @Jawvig

**Comment by GeeSuth** [comment]
from the PRs referenced to this Issue seems @lahma is starting with this once, it's not clear what he faced to make this issue still open until now.

**Comment by lahma** [comment]
As an update, there's no active work from my part. v4 focus is bringing things closer to modern .NET idioms like service collection configuration etc. With landscape changing so fast, I wonder if Blazor would be a good fit or not. Admin functions are considered intranet so it might a worth a shot (unless it becomes Silverlight).

**Comment by Pinox** [comment]
@lahma I think MS for the first time is spot on with Blazor. MS should stay away from any UI frameworks (xaml based). They really suck at it.  Can only sing good praises using Blazor and best of all you can use it with anything from Mudblazor to bootstrap.
